File: backend/app/schemas/game.py
"""Pydantic schemas for games and boxscore data."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class IngestRequest(BaseModel):
    """Request body for ingesting a Sidearm boxscore URL."""

    url: HttpUrl = Field(..., description="Sidearm boxscore page URL")


class TeamStatRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    stat_name: str
    home_value: str | None = None
    away_value: str | None = None
    sort_order: int = 0


class PlayerStatGroupRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    category: str
    team: str | None = None
    columns: list[str]
    rows: list[list[str]]


class ScoringPlayRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    period: str | None = None
    clock: str | None = None
    team: str | None = None
    description: str | None = None
    home_score: int | None = None
    away_score: int | None = None
    sort_order: int = 0


class GameSummary(BaseModel):
    """Lightweight listing view."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    source_url: str
    canonical_uid: str
    source_system: str = "sidearm"
    source_event_id: str | None = None
    sport: str | None = None
    sport_name: str | None = None
    gender: str | None = None
    season: str | None = None
    game_date: str | None = None
    event_shape: str = "team_contest"
    event_status: str = "unknown"
    publish_status: str = "draft"
    home_team: str | None = None
    away_team: str | None = None
    home_score: int | None = None
    away_score: int | None = None
    title: str | None = None
    start_at: datetime | None = None
    end_at: datetime | None = None
    timezone: str | None = None
    location_name: str | None = None
    venue_name: str | None = None
    home_away_neutral: str | None = None
    conference_event: bool = False
    exhibition: bool = False
    first_seen_at: datetime | None = None
    last_seen_at: datetime | None = None
    last_successful_ingest_at: datetime | None = None
    ingested_at: datetime


class EventSourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    source_type: str
    source_url: str
    source_id: str | None = None
    primary_source: bool = False
    discovered_at: datetime
    last_fetched_at: datetime | None = None


class SourceSnapshotRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    parser_version: str
    content_hash: str
    http_status: int | None = None
    fetched_at: datetime


class EventStatusHistoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    from_status: str | None = None
    to_status: str
    reason: str | None = None
    changed_at: datetime


class GameDetail(GameSummary):
    """Full boxscore view."""

    team_stats: list[TeamStatRead] = []
    player_stats: list[PlayerStatGroupRead] = []
    scoring_plays: list[ScoringPlayRead] = []
    event_sources: list[EventSourceRead] = []
    source_snapshots: list[SourceSnapshotRead] = []
    status_history: list[EventStatusHistoryRead] = []
    generated_content: list["GeneratedContentRead"] = []  # noqa: F821


# Imported after GameDetail to avoid circular import at module load time.
from app.schemas.content import GeneratedContentRead  # noqa: E402

GameDetail.model_rebuild()

File: frontend/src/types/game.ts
export interface TeamStat {
  stat_name: string;
  home_value: string | null;
  away_value: string | null;
  sort_order: number;
}

export interface PlayerStatGroup {
  category: string;
  team: string | null;
  columns: string[];
  rows: string[][];
}

export interface ScoringPlay {
  period: string | null;
  clock: string | null;
  team: string | null;
  description: string | null;
  home_score: number | null;
  away_score: number | null;
  sort_order: number;
}

export interface GameSummary {
  id: number;
  source_url: string;
  canonical_uid: string;
  source_system: string;
  source_event_id: string | null;
  sport: string | null;
  sport_name: string | null;
  gender: string | null;
  season: string | null;
  game_date: string | null;
  event_shape: string;
  event_status: string;
  publish_status: string;
  home_team: string | null;
  away_team: string | null;
  home_score: number | null;
  away_score: number | null;
  title: string | null;
  start_at: string | null;
  end_at: string | null;
  timezone: string | null;
  location_name: string | null;
  venue_name: string | null;
  home_away_neutral: string | null;
  conference_event: boolean;
  exhibition: boolean;
  first_seen_at: string | null;
  last_seen_at: string | null;
  last_successful_ingest_at: string | null;
  ingested_at: string;
  last_validated_at: string | null;
}

export interface EventSource {
  source_type: string;
  source_url: string;
  source_id: string | null;
  primary_source: boolean;
  discovered_at: string;
  last_fetched_at: string | null;
}

export interface SourceSnapshot {
  parser_version: string;
  content_hash: string;
  http_status: number | null;
  fetched_at: string;
}

export interface EventStatusHistory {
  from_status: string | null;
  to_status: string;
  reason: string | null;
  changed_at: string;
}

export interface GeneratedContent {
  id: number;
  game_id: number;
  recap: string;
  spotlight_player: string | null;
  spotlight_body: string;
  social_post: string;
  headline: string | null;
  model: string | null;
  generated_at: string;
}

export interface ValidationCheck {
  rule: string;
  passed: boolean;
  detail: string | null;
}

export interface ValidationResult {
  game_id: number;
  passed: boolean;
  publish_status: string;
  checks: ValidationCheck[];
  validated_at: string;
}

export interface PublishEvent {
  id: number;
  game_id: number;
  from_status: string;
  to_status: string;
  trigger: string;
  detail: Record<string, unknown>;
  changed_at: string;
}

import type { AgentRunSummary } from "./agentRun";

export interface GameDetail extends GameSummary {
  team_stats: TeamStat[];
  player_stats: PlayerStatGroup[];
  scoring_plays: ScoringPlay[];
  event_sources: EventSource[];
  source_snapshots: SourceSnapshot[];
  status_history: EventStatusHistory[];
  generated_content: GeneratedContent[];
  agent_runs: AgentRunSummary[];
  publish_events: PublishEvent[];
}
